### ISSUE TITLE
[core/data] pyarrow CSV ParseOptions invalid_row_handler deserialization is faulty

### ISSUE THREAD

**Issue by krfricke** [ray-project/ray#28326]
### What happened + What you expected to happen

First reported here: https://stackoverflow.com/questions/73593586/why-pyarrow-not-detect-invalid-row-handler-when-parsing-param-as-open-csvargs (shorter repro script below).

Using object-store transferred `pyarrow.csv.ParseOptions` drops the `invalid_row_handler`, rendering it unusable.

It's probably best to see this in the reproduction script. We create a CSV file with an invalid row. If the `read_file` command is scheduled as a remote task, it fails with:

```
pyarrow.lib.ArrowInvalid: CSV parse error: Expected 2 columns, got 1: 
```

Script:

```
from pyarrow import csv

import ray


invalid_csv = """f1,f2
3,4
5,6
\x00\x00
7,8"""

source = "test.csv"
with open(source, "w") as f:
    f.write(invalid_csv)


def read_file(path, parse_options):
    # Uncomment this for a fix!
    # parse_options.invalid_row_handler = parse_options.invalid_row_handler

    with open(path, "rb") as f:
        return csv.read_csv(f, parse_options=parse_options)


parse_options = csv.ParseOptions(delimiter=",", invalid_row_handler=lambda i: "skip")


print(read_file(source, parse_options=parse_options))

ray.init()

print(ray.get(ray.remote(read_file).remote(source, parse_options=parse_options)))
```

Oddly enough, re-setting the `invalid_row_handler` resolves the problem. My first intuition here is that pyarrow keeps an internal reference to the invalid row handler, which is deserialized correctly to the python object, but not to the C backend. Re-setting the attribute propagates the correct object to the C backend. Of course I could be mistaken.

Generally this points towards an issue within pyarrow, which I will investigate further. In any case, we can re-set the attribute to resolve the issue for now.

### Versions / Dependencies

Latest master
pyarrow 7.0.0 and 9.0.0

### Reproduction script

See above.

This also affects Ray data, e.g. with this script, which is an abbreviated version of the one in the original issue:

```
from pyarrow import csv

import ray


invalid_csv = """f1,f2
3,4
5,6
\x00\x00
7,8"""

source = "test.csv"
with open(source, "w") as f:
    f.write(invalid_csv)


parse_options = csv.ParseOptions(delimiter=",", invalid_row_handler=lambda i: "skip")

ds = ray.data.read_csv(
    [source],
    parse_options=parse_options,
)

print(ds.count())
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by krfricke** [comment]
Confirmed this is a cloudpickle issue:

```
import cloudpickle
from pyarrow import csv


invalid_csv = """f1,f2
3,4
5,6
\x00\x00
7,8"""

source = "test.csv"
with open(source, "w") as f:
    f.write(invalid_csv)


def read_file(path, parse_options):
    # Uncomment this for a fix!
    # parse_options.invalid_row_handler = parse_options.invalid_row_handler

    with open(path, "rb") as f:
        return csv.read_csv(f, parse_options=parse_options)


parse_options = csv.ParseOptions(delimiter=",", invalid_row_handler=lambda i: "skip")


print(read_file(source, parse_options=parse_options))

parse_options = cloudpickle.loads(cloudpickle.dumps(parse_options))

print(read_file(source, parse_options=parse_options))
```

**Comment by krfricke** [comment]
I believe this is the culprit.

On deser, pyarrow sets the `_invalid_row_handler` attribute

```
    def __setstate__(self, state):
        (self.delimiter, self.quote_char, self.double_quote,
         self.escape_char, self.newlines_in_values,
         self.ignore_empty_lines, self._invalid_row_handler) = state
```

https://github.com/apache/arrow/blob/master/python/pyarrow/_csv.pyx#L555-L558

But it misses the `deref...` call in the setter:

```
    @invalid_row_handler.setter
    def invalid_row_handler(self, value):
        if value is not None and not callable(value):
            raise TypeError("Expected callable or None, "
                            f"got instance of {type(value)!r}")
        self._invalid_row_handler = value
        deref(self.options).invalid_row_handler = MakeInvalidRowHandler(
            <function[PyInvalidRowCallback]> &_handle_invalid_row, value)
```

https://github.com/apache/arrow/blob/master/python/pyarrow/_csv.pyx#L523-L524

I'll file a PR for Ray to circumvent this problem and an issue on the pyarrow side so it can be addressed.

**Comment by krfricke** [comment]
Filed a fix in #28327 and on pyarrow-side in https://github.com/apache/arrow/pull/14061

**Comment by krfricke** [comment]
Just for completeness, there is a workaround for existing installations of Ray (so users don't have to use latest master Ray or wait for the next pyarrow release). The solution is to implement a custom csv datasource and use `ray.data.read_datasource` instead of `ray.data.read_csv`.

Here is the full working example:

```
from pyarrow import csv

import ray


invalid_csv = """f1,f2
3,4
5,6
\x00\x00
7,8"""

source = "test.csv"
with open(source, "w") as f:
    f.write(invalid_csv)


parse_options = csv.ParseOptions(delimiter=",", invalid_row_handler=lambda i: "skip")


class CustomCSVDatasource(ray.data.datasource.CSVDatasource):
    def _read_stream(
        self, f: "pyarrow.NativeFile", path: str, **reader_args
    ):
        import pyarrow
        from pyarrow import csv

        read_options = reader_args.pop(
            "read_options", csv.ReadOptions(use_threads=False)
        )
        parse_options = reader_args.pop("parse_options", csv.ParseOptions())
        # Re-init invalid row handler: https://issues.apache.org/jira/browse/ARROW-17641
        if hasattr(parse_options, "invalid_row_handler"):
            parse_options.invalid_row_handler = parse_options.invalid_row_handler

        reader = csv.open_csv(
            f, read_options=read_options, parse_options=parse_options, **reader_args
        )
        schema = None
        while True:
            try:
                batch = reader.read_next_batch()
                table = pyarrow.Table.from_batches([batch], schema=schema)
                if schema is None:
                    schema = table.schema
                yield table
            except StopIteration:
                return


ds = ray.data.read_datasource(
    CustomCSVDatasource(),
    paths=[source],
    parse_options=parse_options,
)

print(ds.count())
```
